Command hierarchy can execute conditions

MDoerner opened the pull request:
This PR does two things aroud commands.

First, it removes the need to pass a logger to the `CommandBase` contructor, but retains the possibility for the sake of the `DelegateCommand`.

Second, it changes how `EvaluateCanExecute` is handled to the way it is alrady done for commands deriving from `ReactorCommandBase`: derived classes can only add to the `CanEvaluateExecute` evaluation, but not overwrite. This allows to enforce the base classe's can evaluate contract.

This is still WIP because it breaks the `AddTemplateCommand` because this class does not honor its base class's contract; it takes a different type of parameter. This will need to be separated somehow.

I am aware that this PR conflicts heavily with #4833. So, feel free to merge that first. I will then clean up the conflicts. However, this PR might help to actually enforce the can evaluate condition ion subclasses of `ComCommandBase`.

bclothier commented on Rubberduck.Core/UI/Command/CommandBase.cs:
```diff
@@ -12,20 +12,33 @@ public abstract class CommandBase : ICommand
-        protected virtual bool EvaluateCanExecute(object parameter) => true;
         protected abstract void OnExecute(object parameter);
 
+        protected Func<object, bool> CanExecuteEvaluation { get; private set; }
```
This is a nitpick but the name doesn't sound right - it's asking if it can execute an evaluation, which isn't really what we mean. Maybe `EvaluateAllExecuteConditions`?

MDoerner replied:
I will change the name.

bclothier commented on Rubberduck.Core/UI/CodeExplorer/Commands/AddComponentCommand.cs:
```diff
@@ -38,10 +40,9 @@ protected override void OnExecute(object parameter)
         {
-            if (!base.EvaluateCanExecute(parameter) || 
-                !(parameter is CodeExplorerItemViewModel node) ||
+            if (!(parameter is CodeExplorerItemViewModel node) ||
```
I'm inclined to think that this class is doing too much. The primary reason for `AddTemplateCommand` to derive from this class was because that class already has the implementation for creating a VB component in the project. It probably now should be the most derived class (e.g. `AddStdModuleCommand`, `AddClassModuleCommand`, etc.) responsibility to filter for `parameter is CodeExplorerItemViewModel`, or at least push them further out with a new abstract class that provides this specific filtering behavior. 

OTOH, most of the implementations has little to do with the `ICommand` at all; that probably could be factored out into a specialized service for which the derived classes can request and use to create VBComponent as they need.

MDoerner commented:
I could not keep myself from adding some things not directly related to the main aim of this PR:

1. `IRewriteSessionTryRewrite` has been extracted to `IExecutableRewriteSession`, which extends `IRewriteSession`. This allows to pass an `IRewriteSession` around with the guarantee that it does not get executed. (Unless someone casts it.)
2. There is a symptomatic fix for #4897. This issue seems to be caused by the visual tree not being reliable.
3. Fixed false warning that a quick fix could not be applied for refactoring quickfixes. (`TryRewrite` returned false if nothing was to be rewritten. That is reversed now, i.e. it will return `true` if no rewriter has been checked out.)
4. Refactorings now communicate a failure to rewrite by throwing a `RewriteFailedException`, which is used by the `QuickFixProvider` to trigger the failure notification.

MDoerner commented:
I have extracted adding components out of the component `AddComponentCommend` and changed the version adding code with attributes to actually have a chance to work for VB6. However, I am not completely sure whether it works. (I cannot test this myself.)

Moreover, I made the templates actually work when the UI language is soemthing else than English.

bclothier commented on Rubberduck.Core/UI/CodeExplorer/CodeExplorerAddComponentService.cs:
```diff
@@ -0,0 +1,122 @@
+            using (var hostApp = _vbe.HostApplication())
+            {
+                return hostApp?.ApplicationName == "Access" 
+                    ? "Option Compare Database" 
```
Referencing #4628 -- this probably ought to be cleaned up when a PR is opened to address that issue.

bclothier commented on Rubberduck.Core/UI/CodeExplorer/Commands/AddTemplateCommand.cs:
```diff
@@ -59,6 +66,25 @@ private bool SpecialEvaluateCanExecute(object parameter)
+            try
+            {
+                var project = node.Declaration.Project;
+                return AllowableProjectTypes.Contains(project.Type);
```
Shouldn't the `Project` be wrapped in a `using`?

MDoerner replied:
No, the project on declarations is ultimately managed by the projects repository. Disposing it will break things.

bclothier commented on Rubberduck.Core/UI/CodeExplorer/Commands/AddTemplateCommand.cs:
```diff
@@ -59,6 +66,25 @@ private bool SpecialEvaluateCanExecute(object parameter)
+            }
+            catch (COMException)
+            {
+                return false;
```
I'm inclined to think that it should at least get logged to `Debug` or `Trace` rather than silently swallowing.

MDoerner replied:
Will log ad `Debug`.

MDoerner replied:
I would like to revert my position. The entire `try` should go. Handling exceptions by logging them is already done in `CommandBase`.

bclothier commented on Rubberduck.Core/UI/CodeExplorer/Commands/AddTemplateCommand.cs:
```diff
@@ -4,11 +4,12 @@
 namespace Rubberduck.UI.CodeExplorer.Commands
 {
-    public class AddTemplateCommand : CodeExplorerCommandBase
+    public class AddTemplateCommand : CommandBase
```
TBH, it feels a bit weird to make it a generic Command when it's a part of the code explorer. 

I think it can be made compatible with the `CodeExplorerCommandBase` by changing `ApplicableNodes` to something else and have it contain only one member for the tuple type. The additional conditions from the original `ApplicableNodes` (renamed to something else) can be then added instead.

MDoerner replied:
This will require some investigation with the debugger. After all the _Node_ in `ApplicableNodeTypes` is not really true. I will just have to find out the appropriate type and maybe change slightly how the check works.

bclothier commented on Rubberduck.Core/UI/CodeExplorer/Commands/AddTemplateCommand.cs:
```diff
@@ -42,7 +42,8 @@ public AddTemplateCommand(ICodeExplorerAddComponentService addComponentService,
 
-        public ComponentType ComponentType => ComponentType.Undefined;
+        //We need a valid component type to add the component in the first place. Then the module content gets overwritten.
+        public ComponentType ComponentType => ComponentType.ClassModule;
```
I think we need to pass it from the resource or something; it might not necessarily be a class module in future. At least change the comment to a `TODO` or `FIXME` so it's marked.

MDoerner replied:
For VBA, the `ComponentType` used to create the component is entirely irrelevant bacause changing the contents via the file will change it to the appropriate one.

However, I am not sure whether this works for VB6. It would be good if somebody who has the IDE could test whether this works for templates with a different component type.

Vogel612 commented on Rubberduck.Core/UI/CodeExplorer/CodeExplorerAddComponentService.cs:
```diff
@@ -0,0 +1,122 @@
+
+            var prefixInModule = additionalPrefixInModule != null
+                ? $"{basePrefixInModule}{Environment.NewLine}{additionalPrefixInModule}"
+                : basePrefixInModule;
```
StringBuilder should be able to simplify this a bit...

```
var modulePrefix = new StringBuilder(folderAnnotation);
if (optionCompare != null) {
  modulePrefix.Append(Environment.NewLine).Append(optionCompare);
}
if (additionalPrefixInModule != null) {
  modulePrefix.Append(Environment.NewLine).Append(additionalPrefixInModule);
}
...
```

Vogel612 commented on Rubberduck.Core/UI/CodeExplorer/CodeExplorerAddComponentService.cs:
```diff
@@ -0,0 +1,122 @@
+            var declaration = node?.Declaration;
+            if (declaration == null)
+            {
+                return ActiveProjectName();
```
Shouldn't this be `ActiveProjectFolder`?

MDoerner replied:
Yes, it should.

Vogel612 commented on Rubberduck.Core/UI/CodeExplorer/Commands/CodeExplorerFindAllReferencesCommand.cs:
```diff
@@ -24,6 +24,15 @@ public CodeExplorerFindAllReferencesCommand(RubberduckParserState state, FindAll
+        private bool SpecialEvaluateCanExecute(object parameter)
+        {
+            return ((ICodeExplorerNode)parameter).Declaration != null &&
+                   (!(parameter is CodeExplorerReferenceViewModel reference) || !reference.IsDimmed) &&
```
could we de-morgan this into `!(parameter is CodeExplorereReferenceViewModel reference && reference.IsDimmed)`? Or am I missing something because I can't think straight?

Vogel612 replied:
Also I think we should pull the parser-state check to an earlier position, that is the cheapest check and it probably applies a lot...

MDoerner replied:
TBH, I have blindly copied from the original `EvaluateCanExecute`.

I will adjust this.
